File: android/src/main/kotlin/dev/brewkits/native_workmanager/store/TaskStore.kt
package dev.brewkits.native_workmanager.store

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.util.Log

/**
 * Persistent task storage for Android using SQLite.
 *
 * Uses built-in Android SQLite Framework (no third-party dependencies).
 * Features:
 * - WAL (Write-Ahead Logging) mode for concurrent read/write.
 * - Atomic transactions for data integrity.
 * - Automatic sensitive data redaction (Sanitization).
 * - "Zombie" task recovery (resets tasks stuck in 'running' state).
 */
internal class TaskStore(context: Context) {

    data class TaskRecord(
        val taskId: String,
        val tag: String?,
        val status: String,
        val workerClassName: String,
        val workerConfig: String?,
        val createdAt: Long,
        val updatedAt: Long,
        val resultData: String?,
        val constraintsJson: String? = null,
        val lastProgressJson: String? = null
    )

    private val dbHelper = DatabaseHelper.getInstance(context)

    /**
     * Recover "Zombie" tasks that are stuck in 'running' state.
     * 
     * FIX #03: Added Heartbeat check. Only recover tasks that have been in 'running'
     * state without any update for more than 5 minutes. This prevents killing
     * tasks that were just started by the OS during app launch.
     */
    fun recoverZombieTasks() {
        val now = System.currentTimeMillis()
        val timeoutMs = 5 * 60 * 1000L // 5 minutes heartbeat timeout
        val threshold = now - timeoutMs

        val cv = ContentValues().apply {
            put("status", "failed")
            put("result_data", "{\"message\": \"Process died or hung (heartbeat timeout)\", \"shouldRetry\": true}")
            put("updated_at", now)
        }
        
        // Only update where status is running AND updated_at is older than 5 minutes
        val count = dbHelper.writableDatabase.update(
            "tasks", 
            cv, 
            "status = 'running' AND updated_at < ?", 
            arrayOf(threshold.toString())
        )
        
        if (count > 0) {
            Log.d("TaskStore", "Recovered $count zombie tasks (heartbeat based)")
        }
    }

    fun upsert(
        taskId: String,
        tag: String?,
        status: String,
        workerClassName: String,
        workerConfig: String?,
        constraintsJson: String? = null
    ) {
        val now = System.currentTimeMillis()
        val db = dbHelper.writableDatabase
        
        // Sanitize config before persisting to prevent token leakage.
        val sanitizedConfig = sanitizeConfig(workerConfig)
        
        db.beginTransaction()
        try {
            val existingCreatedAt = db.rawQuery(
                "SELECT created_at FROM tasks WHERE task_id = ?", arrayOf(taskId)
            ).use { c -> if (c.moveToFirst()) c.getLong(0) else now }

            val cv = ContentValues().apply {
                put("task_id", taskId)
                put("tag", tag)
                put("status", status)
                put("worker_class", workerClassName)
                put("worker_config", sanitizedConfig)
                put("created_at", existingCreatedAt)
                put("updated_at", now)
                put("constraints_json", constraintsJson)
            }
            db.insertWithOnConflict("tasks", null, cv, SQLiteDatabase.CONFLICT_REPLACE)
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }
    }

    fun updateStatus(taskId: String, status: String, resultData: String? = null) {
        Log.d("TaskStore", "Updating task $taskId status to $status, resultData length: ${resultData?.length ?: 0}")
        val cv = ContentValues().apply {
            put("status", status)
            put("updated_at", System.currentTimeMillis())
            if (resultData != null) put("result_data", resultData)
        }
        dbHelper.writableDatabase.update("tasks", cv, "task_id = ?", arrayOf(taskId))
    }

    fun updateProgress(taskId: String, progressJson: String) {
        val cv = ContentValues().apply {
            put("last_progress_json", progressJson)
            put("updated_at", System.currentTimeMillis())
        }
        dbHelper.writableDatabase.update("tasks", cv, "task_id = ?", arrayOf(taskId))
    }

    fun getTask(taskId: String): TaskRecord? =
        dbHelper.readableDatabase.rawQuery("SELECT * FROM tasks WHERE task_id = ?", arrayOf(taskId))
            .use { c -> if (c.moveToFirst()) c.toRecord() else null }

    fun getTasksByTag(tag: String): List<TaskRecord> =
        dbHelper.readableDatabase.rawQuery("SELECT * FROM tasks WHERE tag = ? ORDER BY updated_at DESC", arrayOf(tag))
            .use { c ->
                val list = mutableListOf<TaskRecord>()
                while (c.moveToNext()) list.add(c.toRecord())
                list
            }

    fun getAllTasks(): List<TaskRecord> =
        dbHelper.readableDatabase.rawQuery("SELECT * FROM tasks ORDER BY updated_at DESC", null)
            .use { c ->
                val list = mutableListOf<TaskRecord>()
                while (c.moveToNext()) list.add(c.toRecord())
                list
            }

    fun delete(taskId: String) {
        dbHelper.writableDatabase.delete("tasks", "task_id = ?", arrayOf(taskId))
    }

    fun deleteCompleted(olderThanMs: Long = 0L, batchSize: Int = 1000) {
        val threshold = if (olderThanMs > 0) System.currentTimeMillis() - olderThanMs else Long.MAX_VALUE
        // Batch delete to avoid holding a long write-lock that blocks concurrent upsert calls.
        // Repeat until fewer rows than batchSize are deleted (i.e. no more rows to clean).
        do {
            val deleted = dbHelper.writableDatabase.delete(
                "tasks",
                "task_id IN (SELECT task_id FROM tasks WHERE status IN ('completed','failed','cancelled') AND updated_at < ? LIMIT ?)",
                arrayOf(threshold.toString(), batchSize.toString())
            )
            if (deleted < batchSize) break
        } while (true)
    }

    private fun android.database.Cursor.toRecord(): TaskRecord {
        val constraintsColIdx = getColumnIndex("constraints_json")
        val progressColIdx = getColumnIndex("last_progress_json")
        return TaskRecord(
            taskId          = getString(getColumnIndexOrThrow("task_id")),
            tag             = getString(getColumnIndexOrThrow("tag")),
            status          = getString(getColumnIndexOrThrow("status")),
            workerClassName = getString(getColumnIndexOrThrow("worker_class")),
            workerConfig    = getString(getColumnIndexOrThrow("worker_config")),
            createdAt       = getLong(getColumnIndexOrThrow("created_at")),
            updatedAt       = getLong(getColumnIndexOrThrow("updated_at")),
            resultData      = getString(getColumnIndexOrThrow("result_data")),
            constraintsJson = if (constraintsColIdx >= 0) getString(constraintsColIdx) else null,
            lastProgressJson = if (progressColIdx >= 0) getString(progressColIdx) else null
        )
    }

    companion object {
        private val SENSITIVE_KEYS = setOf(
            "authToken", "authorization", "cookies", "password", "secret",
            "accessToken", "refreshToken", "apiKey", "token", "bearer"
        )

        fun sanitizeConfig(json: String?): String? {
            if (json == null) return null
            return try {
                val obj = org.json.JSONObject(json)
                sanitizeObject(obj)
                obj.toString()
            } catch (_: Exception) { json }
        }

        private fun sanitizeObject(obj: org.json.JSONObject) {
            val keys = obj.keys().asSequence().toList()
            for (key in keys) {
                if (SENSITIVE_KEYS.any { it.equals(key, ignoreCase = true) }) {
                    obj.put(key, "[REDACTED]")
                } else {
                    when (val value = obj.opt(key)) {
                        is org.json.JSONObject -> sanitizeObject(value)
                        is org.json.JSONArray  -> sanitizeArray(value)
                        else -> { }
                    }
                }
            }
        }

        private fun sanitizeArray(arr: org.json.JSONArray) {
            for (i in 0 until arr.length()) {
                when (val item = arr.opt(i)) {
                    is org.json.JSONObject -> sanitizeObject(item)
                    is org.json.JSONArray  -> sanitizeArray(item)
                    else -> { }
                }
            }
        }
    }

    fun getActiveTaskCount(): Int {
        val db = dbHelper.readableDatabase
        db.rawQuery("SELECT COUNT(*) FROM tasks WHERE status IN ('pending', 'running')", null).use { cursor ->
            if (cursor.moveToFirst()) return cursor.getInt(0)
        }
        return 0
    }

    fun getFailedTaskCount(): Int {
        val db = dbHelper.readableDatabase
        db.rawQuery("SELECT COUNT(*) FROM tasks WHERE status = 'failed'", null).use { cursor ->
            if (cursor.moveToFirst()) return cursor.getInt(0)
        }
        return 0
    }

    fun getCompletedTaskCount(): Int {
        val db = dbHelper.readableDatabase
        db.rawQuery("SELECT COUNT(*) FROM tasks WHERE status = 'success'", null).use { cursor ->
            if (cursor.moveToFirst()) return cursor.getInt(0)
        }
        return 0
    }

    fun TaskRecord.toFlutterMap(): Map<String, Any?> = mapOf(
        "taskId"          to taskId,
        "tag"             to tag,
        "status"          to status,
        "workerClassName" to workerClassName,
        "workerConfig"    to workerConfig,
        "createdAt"       to createdAt,
        "updatedAt"       to updatedAt,
        "resultData"      to try {
            resultData?.let { org.json.JSONObject(it).toMap() }
        } catch (_: Exception) {
            null
        }
    )

    private fun org.json.JSONObject.toMap(): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>()
        val keys = keys()
        while (keys.hasNext()) {
            val key = keys.next()
            var value = get(key)
            if (value is org.json.JSONArray) {
                value = value.toList()
            } else if (value is org.json.JSONObject) {
                value = value.toMap()
            }
            if (value == org.json.JSONObject.NULL) value = null
            map[key] = value
        }
        return map
    }

    private fun org.json.JSONArray.toList(): List<Any?> {
        val list = mutableListOf<Any?>()
        for (i in 0 until length()) {
            var value = get(i)
            if (value is org.json.JSONArray) {
                value = value.toList()
            } else if (value is org.json.JSONObject) {
                value = value.toMap()
            }
            if (value == org.json.JSONObject.NULL) value = null
            list.add(value)
        }
        return list
    }
}

File: ios/native_workmanager/Sources/native_workmanager/utils/TaskStore.swift
import Foundation

/// Persistent task storage for iOS using atomic file-based JSON records.
/// 
/// Instead of a database, each task is stored as a separate .json file in the
/// app's Application Support directory. This ensures high reliability and
/// performance without the overhead of SQLite or CoreData.
///
/// Uses .atomic writing to guarantee data integrity across app crashes and reboots.
@available(iOS 13.0, *)
public class TaskStore {
    
    public struct TaskRecord: Codable {
        public let taskId: String
        public let tag: String?
        public let status: String
        public let workerClassName: String
        public let workerConfig: String?
        public let createdAt: Int64
        public let updatedAt: Int64
        public let resultData: String?
        public let constraintsJson: String?
        
        /// Returns a sanitized version of the record with sensitive config fields removed.
        /// Prevents leaking auth tokens or passwords via backups or file system dumps.
        public func sanitized() -> TaskRecord {
            guard let config = workerConfig else { return self }
            return TaskRecord(
                taskId: taskId,
                tag: tag,
                status: status,
                workerClassName: workerClassName,
                workerConfig: TaskStore.sanitizeConfig(config),
                createdAt: createdAt,
                updatedAt: updatedAt,
                resultData: resultData,
                constraintsJson: constraintsJson
            )
        }
        
        /// For backward compatibility with existing plugin logic.
        public func toFlutterMap() -> [String: Any?] {
            return [
                "taskId": taskId,
                "tag": tag,
                "status": status,
                "workerClassName": workerClassName,
                "createdAt": createdAt,
                "updatedAt": updatedAt,
                "resultData": resultData
            ]
        }
    }
    
    public static let shared = TaskStore()
    
    private let fileManager = FileManager.default
    private let tasksDir: URL
    
    private init() {
        // Use Application Support directory for persistent data (standard iOS practice).
        let appSupport = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        tasksDir = appSupport.appendingPathComponent("native_workmanager/tasks", isDirectory: true)
        
        // Ensure the directory exists.
        if !fileManager.fileExists(atPath: tasksDir.path) {
            try? fileManager.createDirectory(at: tasksDir, withIntermediateDirectories: true)
        }
    }
    
    /// Persist or update a task record.
    /// Uses .atomic writing to prevent file corruption.
    public func upsert(
        taskId: String,
        tag: String?,
        status: String,
        workerClassName: String,
        workerConfig: String?,
        constraintsJson: String? = nil
    ) {
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let existing = getTask(taskId: taskId)
        
        let record = TaskRecord(
            taskId: taskId,
            tag: tag,
            status: status,
            workerClassName: workerClassName,
            workerConfig: workerConfig,
            createdAt: existing?.createdAt ?? now,
            updatedAt: now,
            resultData: existing?.resultData,
            constraintsJson: constraintsJson
        )
        
        save(record)
    }
    
    /// Update status and optional result data for an existing task.
    public func updateStatus(taskId: String, status: String, resultData: String? = nil) {
        guard var record = getTask(taskId: taskId) else { return }
        
        record = TaskRecord(
            taskId: record.taskId,
            tag: record.tag,
            status: status,
            workerClassName: record.workerClassName,
            workerConfig: record.workerConfig,
            createdAt: record.createdAt,
            updatedAt: Int64(Date().timeIntervalSince1970 * 1000),
            resultData: resultData ?? record.resultData,
            constraintsJson: record.constraintsJson
        )
        
        save(record)
    }
    
    /// Retrieve a task by ID.
    public func getTask(taskId: String) -> TaskRecord? {
        let fileURL = tasksDir.appendingPathComponent("\(taskId).json")
        guard let data = try? Data(contentsOf: fileURL) else { return nil }
        return try? JSONDecoder().decode(TaskRecord.self, from: data)
    }
    
    /// Compatibility alias for getTask
    public func task(taskId: String) -> TaskRecord? {
        return getTask(taskId: taskId)
    }
    
    /// Retrieve all tasks, sorted by updatedAt descending.
    public func getAllTasks() -> [TaskRecord] {
        guard let files = try? fileManager.contentsOfDirectory(at: tasksDir, includingPropertiesForKeys: nil) else {
            return []
        }
        
        let records = files.compactMap { fileURL -> TaskRecord? in
            guard fileURL.pathExtension == "json" else { return nil }
            guard let data = try? Data(contentsOf: fileURL) else { return nil }
            return try? JSONDecoder().decode(TaskRecord.self, from: data)
        }
        
        return records.sorted { $0.updatedAt > $1.updatedAt }
    }
    
    /// Compatibility alias for getAllTasks
    public func allTasks() -> [TaskRecord] {
        return getAllTasks()
    }
    
    /// Delete a task record.
    public func delete(taskId: String) {
        let fileURL = tasksDir.appendingPathComponent("\(taskId).json")
        try? fileManager.removeItem(at: fileURL)
    }

    /// Recover "Zombie" tasks that are stuck in 'running' state.
    /// 
    /// FIX #03: Added Heartbeat check. Only recover tasks that have been in 'running'
    /// state without any update for more than 5 minutes. This prevents killing
    /// tasks that were just started by the OS during app launch.
    public func recoverZombieTasks() {
        let all = getAllTasks()
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let timeoutMs: Int64 = 5 * 60 * 1000 // 5 minutes heartbeat timeout
        
        var recoveredCount = 0
        for record in all where record.status == "running" {
            if (now - record.updatedAt) > timeoutMs {
                updateStatus(
                    taskId: record.taskId,
                    status: "failed",
                    resultData: "{\"message\": \"Process died or hung (heartbeat timeout)\", \"shouldRetry\": true}"
                )
                recoveredCount += 1
            }
        }
        if recoveredCount > 0 {
            print("TaskStore: Recovered \(recoveredCount) zombie tasks (heartbeat based)")
        }
    }

    /// Delete old tasks (cleanup).
    public func deleteCompleted(olderThanMs: Int64 = 0) {
        let threshold = olderThanMs > 0 ? Int64(Date().timeIntervalSince1970 * 1000) - olderThanMs : Int64.max
        let all = getAllTasks()
        
        for record in all {
            let s = record.status
            if (s == "completed" || s == "failed" || s == "cancelled") && record.updatedAt < threshold {
                delete(taskId: record.taskId)
            }
        }
    }
    
    private func save(_ record: TaskRecord) {
        let fileURL = tasksDir.appendingPathComponent("\(record.taskId).json")
        
        // Sanitize before persisting to prevent sensitive data leakage.
        let sanitized = record.sanitized()
        
        if let data = try? JSONEncoder().encode(sanitized) {
            // .atomic writing ensures the old file is only replaced if the new write succeeds.
            try? data.write(to: fileURL, options: .atomic)
        }
    }
    
    // MARK: - Security & Sanitization
    
    private static let sensitiveKeys = [
        "authToken", "authorization", "cookies", "password", "secret",
        "accessToken", "refreshToken", "apiKey", "token", "bearer"
    ]
    
    /// Recursively strips sensitive keys from JSON config before persistence.
    public static func sanitizeConfig(_ json: String) -> String {
        guard let data = json.data(using: .utf8),
              var dict = (try? JSONSerialization.jsonObject(with: data) as? [String: Any]) else {
            return json
        }
        
        sanitizeDictionary(&dict)
        
        if let sanitizedData = try? JSONSerialization.data(withJSONObject: dict, options: []),
           let result = String(data: sanitizedData, encoding: .utf8) {
            return result
        }
        return json
    }
    
    public static func sanitizeDictionary(_ dict: inout [String: Any]) {
        for key in dict.keys {
            if sensitiveKeys.contains(where: { $0.lowercased() == key.lowercased() }) {
                dict[key] = "[REDACTED]"
            } else if var nestedDict = dict[key] as? [String: Any] {
                sanitizeDictionary(&nestedDict)
                dict[key] = nestedDict
            } else if var nestedArray = dict[key] as? [[String: Any]] {
                for i in 0..<nestedArray.count {
                    sanitizeDictionary(&nestedArray[i])
                }
                dict[key] = nestedArray
            }
        }
    }
}
